### DISCUSSION TITLE
For the love of God : introduce mouse mode for zooming and moving the canvas

### DISCUSSION THREAD

**Discussion by melyaaka** [excalidraw/excalidraw#5515] [General]
As with all apps that use a large amount of space to add and place objects, the brain thinks about them as a sort of a map. 
and to navigate a map, there is nothing simpler than : 
- scroll wheel for zoom in / out 
- drag with right button to move the canva

Why on earth do we need to press a modifier key for this behavior to take place ? when in reality it's the better and more useful way. 

99% of the time that I want to use the scroll wheel, what I what to do is change the scale in order to explore other areas of the canvas, not to move it up and down. the current way of doing things moves the canvas up and down which is rarely what the user wants. also with the current way of doing things the scroll wheel cannot be used to move the canva horizontally... 

Google maps uses the scroll wheel to zoom/in and out. it would suck if it was using it to move the map slightly up or down, and it would take 100 times more effort to explore another place of the map if it worked that way. 

Miro uses scroll wheel for zoom in and out. 

Even the graph view in obsidian works this way. (scroll wheel for zoom)

This is how it is done in all apps that know a thing or two about user friendly interfaces. 

So I have no idea why excalidraw has to be the way it is. At the very least you can offer a trackpad mode (the current one) and a mouse mode (with the changes that I proposed at the beginning). 

I quit using a mindmap app called ithoughts for this exact reason. I would have HUUUGE maps, and when I wanted to explore them the scrolling wheel just moves the entire mind map slightly up or slightly down instead of allowing me to quickly change perspective. It was perfect in all other ways. But the navigation part was a dealbreaker.

**Comment by sojusnik** [comment]
You can click and hold the mouse wheel button and drag the canvas as well, but a mouse wheel button click also inserts a previously copied text on my OS (Ubuntu on Brave Browser), so this solution isn't optimal at the moment, because it often inserts text, while I just wanted to move the canvas. I wish that Excalidraw would block pasting after a middle mouse click.

**Comment by sean1832** [comment]
This will be a very nice addition to the Excalidraw, as it will drastically improve productivity for people like me who is used to zoom with scroll wheel. Perhaps implement a zoom mode, and giving user a choice to switch between.

**Reply by Alchete** [reply]
How do you drag with a right-click to move the canvas???  This is not standard.  Right-click in any app is ALWAYS a menu.

I just discovered this web app today.  I think the devs have done an absolutely phenomenal job with this UI.  One of the best I've ever seen.  They copied every applicable Photoshop shortcut key, and they all work here as expected.  The UI elements are fluid and intuitive.  It's minimalist.  It's well-documented.  

- Middle mousewheel + drag moves the canvas, exactly as I would have expected.  It can't be a plain left or right click for obvious reasons.

- Alternatively, spacebar+left-click also moves the canvas, exactly as it does in Photoshop.  This is referred to as a "spring-loaded mode".

- CTRL-mousewheel zooms, as it does in a browser.  They also added a "Google Maps" zoom-to-cursor effect, just as I hoped.  But yes, here I can see a strong argument to just use the plain mouse wheel to zoom, as vertical scrolling doesn't seem to have much use.

Honestly, my biggest complaint is that they don't have a bright yellowish color in their palette, which is my go-to for dark backgrounds. :)

**Comment by papucho** [comment]
Please do add this option. It would be amazing!

**Reply by juhasuni** [reply]
> This is how it is done in all apps that know a thing or two about user friendly interfaces.

That's not true actually. The current behavior is also how e.g. Figma and Tldraw works (CMD+Scroll to zoom).

**Reply by papucho** [reply]
Miro and Figma support scroll to zoom and right-click panning. I consider that to be just amazing.

The right-click is used to pan when hold but displays the context menu if is just clicked (without holding).

**Comment by alastair-games** [comment]
Please allow mouse wheel to zoom in and out, its how it works on other apps I use. 

Could just be a setting to disable having to hold "control". 
Since I'll never need to use the vertical scrolling function, when I can move the canvas with middle click anyway.

It's significantly more convenient to be able to navigate and zoom with one hand, rather than being forced to use both hands to navigate.

Thank you!

**Comment by arch1en** [comment]
Bump.

There could be an option to toggle between current solution and mouse scroll zoom in/out.
Mouse scroll is way more handy for whiteboard browsing.

Cheers!

**Comment by Frzjck** [comment]
Wish this was default behaviour but a chrome extension would also do...

**Comment by sricks** [comment]
Created this violentmonkey script with AI. Sets right click to pan, scroll wheel to zoom. Works well for me in brave. Right click menu in excalidraw is not really useful so I don't lose anything.

```js
// ==UserScript==
// @name         Excalidraw: Wheel-Zoom + Right-Click Pan
// @namespace    excalidraw-wheel-zoom-rmb-pan
// @version      1.0
// @description  Scroll wheel zooms (centered on cursor). Right-click + drag pans. Native context menu suppressed.
// @match        https://excalidraw.com/*
// @grant        none
// @run-at       document-idle
// ==/UserScript==

(function () {
  'use strict';

  const mkOpts = (e, extra) => Object.assign({
    bubbles: true, cancelable: true, composed: true, view: window,
    clientX: e.clientX, clientY: e.clientY,
    screenX: e.screenX, screenY: e.screenY,
    ctrlKey: e.ctrlKey, shiftKey: e.shiftKey, altKey: e.altKey, metaKey: e.metaKey,
  }, extra);

  const firePointer = (type, target, e, extra) =>
    target.dispatchEvent(new PointerEvent(type, mkOpts(e, Object.assign({
      pointerId: e.pointerId ?? 1, pointerType: 'mouse', isPrimary: true,
    }, extra))));

  const fireMouse = (type, target, e, extra) =>
    target.dispatchEvent(new MouseEvent(type, mkOpts(e, extra)));

  // ---------- 1. Wheel -> Zoom (centered on cursor) ----------
  window.addEventListener('wheel', (e) => {
    if (!e.isTrusted || e.ctrlKey || e.metaKey) return; // ignore synthetic + already-zoom events
    e.preventDefault();
    e.stopImmediatePropagation();

    e.target.dispatchEvent(new WheelEvent('wheel', mkOpts(e, {
      deltaX: 0, deltaY: e.deltaY, deltaZ: e.deltaZ, deltaMode: e.deltaMode,
      ctrlKey: true, // <- Excalidraw's native zoom trigger
    })));
  }, { capture: true, passive: false });

  // ---------- 2. Right-click + drag -> Pan ----------
  let panning = false;
  let panTarget = null;

  window.addEventListener('contextmenu', (e) => {
    if (!e.isTrusted) return;
    e.preventDefault();
    e.stopImmediatePropagation();
  }, { capture: true });

  window.addEventListener('pointerdown', (e) => {
    if (!e.isTrusted || e.button !== 2) return;
    e.preventDefault();
    e.stopImmediatePropagation();
    panning = true;
    panTarget = e.target;
    firePointer('pointerdown', panTarget, e, { button: 1, buttons: 4 });
    fireMouse('mousedown', panTarget, e, { button: 1, buttons: 4 });
  }, { capture: true });

  window.addEventListener('pointermove', (e) => {
    if (!e.isTrusted || !panning) return;
    if (!(e.buttons & 2)) { endPan(e); return; } // safety: right button released elsewhere
    e.preventDefault();
    e.stopImmediatePropagation();
    firePointer('pointermove', panTarget, e, { button: -1, buttons: 4 });
    fireMouse('mousemove', panTarget, e, { button: -1, buttons: 4 });
  }, { capture: true });

  function endPan(e) {
    if (!panning) return;
    panning = false;
    e.preventDefault();
    e.stopImmediatePropagation();
    firePointer('pointerup', panTarget, e, { button: 1, buttons: 0 });
    fireMouse('mouseup', panTarget, e, { button: 1, buttons: 0 });
    panTarget = null;
  }

  window.addEventListener('pointerup', (e) => {
    if (!e.isTrusted || e.button !== 2) return;
    endPan(e);
  }, { capture: true });

  window.addEventListener('blur', () => { panning = false; panTarget = null; });
})();
```
